topiary/src/atom_collection.rs

```diff
@@ -266,23 +269,25 @@ impl AtomCollection {
     pub fn apply_prepends_and_appends(&mut self) {
         let mut expanded: Vec<Atom> = Vec::new();
 
-        for atom in self.atoms.iter() {
+        for atom in self.atoms.iter_mut() {
             match atom {
                 Atom::Leaf { id, .. } => {
-                    for prepended in self.prepend.entry(*id).or_default() {
-                        expanded.push(prepended.clone());
-                    }
+                    let prepends = self.prepend.entry(*id).or_default();
+                    let appends = self.append.entry(*id).or_default();
 
-                    expanded.push(atom.clone());
+                    // Rather than cloning the atom from the old vector, we
+                    // swap it out with an empty one.
+                    let mut swapped_atom = Atom::Empty;
+                    mem::swap(&mut swapped_atom, atom);
 
-                    for appended in self.append.entry(*id).or_default() {
-                        log::debug!("Applying append of {appended:?} to {atom:?}.");
-                        expanded.push(appended.clone());
-                    }
+                    expanded.append(prepends);
+                    expanded.push(swapped_atom);
+                    expanded.append(appends);
                 }
                 _ => {
-                    log::debug!("Not a leaf: {atom:?}");
-                    expanded.push(atom.clone());
+                    let mut swapped_atom = Atom::Empty;
+                    mem::swap(&mut swapped_atom, atom);
+                    expanded.push(swapped_atom);
```

Comment:
This swapping business is a bit harder to understand as a Rust newbie. AIUI, it's similar to what `Option::take` is doing to satisfy ownership. If it's an idiomatic thing to do -- which it probably is -- then I think it's fine; if not, _maybe_ it deserves a bit more comment.

(Incidentally, you removed the `log::debug!`. Was that intentional?

Reply:
Added more explanation. And added back the logging.